### PR TITLE
Analytic sphere-box, sphere-cylinder, and capsule-cylinder collisions

### LINKED ISSUE #57476: [GodotPhysics] CharacterBody3D gets stuck on flat ends of static cylinders
### Godot version

4.0

### System information

Windows 10

### Issue description

Moving a CharacterBody3D along the flat end of a cylinder does not behave the same as sliding along a flat wall. The CharacterBody3D can even become stuck in place and unable to move away form the cylinder once it reaches the edges of the cylinder's flat surface
CylinderIssue
.

### Steps to reproduce

- Place a CharacterBody3D on a flat surface
- Place a cylindrical StaticBody3D collider near the CharacterBody3D, and rotate it 90° along the X axis so the bottom or top of the cylinder becomes a wall.
- In play mode move the character so that it collides with the flat end of the cylinder, and try to slide along it diagonally like a wall.

(For some reason if the cylinder has its scale set to (2.8, 2, 2.8) and is placed about 1 unit above the surface of the floor the issue is more present)

### Minimal reproduction project

[CharacterAndCylinderIssue.zip](https://github.com/godotengine/godot/files/7968362/CharacterAndCylinderIssue.zip)

### LINKED ISSUE #70405: [GodotPhysics] Extremely odd Rigid capsule with Static cylinder collision
### Godot version

4.0 beta 9

### System information

Manjaro 6.0.14

### Issue description

(3D) Rigid bodies with capsule shapes seem to have completely broken collision with static bodies with cylinder shapes, as they collide as if they were a cylinder of the same total height, and on top of that they don't topple at all when their centre of mass is over an edge.
image

### Steps to reproduce

Start a new project. In the 3D scene add a StaticBody3D with a cylinder collision shape and a RigidBody3D with a capsule collision shape. (Add a camera and meshes to see what's going on.) Move the RigidBody3D above the StaticBody3D so its centre of mass' downward projection doesn't hit the cylinder. Play the scene

For the minimal reproduction project: Play the scene

### Minimal reproduction project

[bug display.zip](https://github.com/godotengine/godot/files/10279550/bug.display.zip)

### PR DESCRIPTION
**Pull request by peastman**
This addresses part of #70615.  It replaces the SAT code for three collision types with analytic solvers.  This tends to be faster and more accurate.

*Physics squad edit:*
- Fixes https://github.com/godotengine/godot/issues/70405
- Fixes https://github.com/godotengine/godot/issues/57476

### CODE REVIEW

**Comment by Malcolmnixon** [comment]
My testing shows:
 - sphere/box collision detection appears to work fine.
 - sphere/cylinder collision detection appears to work fine.
 - capsule/cylinder collision detection appears to work fine.

I did notice an unrelated effect however - when I move_and_collide the objects I'm getting different Z positions which makes it look like the objects are bouncing slightly. The position distance is only 0.015625 and I think it's caused by the GodotSpace3D::test_body_motion iterative walking loop; however it may explain why the Godot 4 VR games feel like the players height is bouncing up and down slightly.

**Comment by Malcolmnixon** [comment]
As requested by @fabriceci in the physics contributors chat; I've started work on an automated test suite for physics collisions. It currently only has the following configurations:
 - sphere/sphere
 - sphere/cylinder
 - sphere/box
 - sphere/capsule
 - capsule/capsule

Running against Godot 4 beta 10 shows 9 test failures including capsule/capsule, sphere/capsule, sphere/cylinder
Running against #70660 shows the capsule/capsule and sphere/capsule is fixed
Running against #70787 shows the sphere/cylinder is fixed

I'll add more tests including the capsule/cylinder shortly.

**Comment by peastman** [comment]
That's excellent!  I've had it in mind for a while to write an automated test that would do the following.

- Loop over every possible combination of two primitive types.
- For each one, place them at a series of random positions and orientations and check for collisions.
- Repeat the check, replacing either or both objects with a high resolution mesh.
- Make sure they agree up to the accuracy of the mesh.

I've been waiting for #67481 to get merged first, since right now collisions on large meshes are too slow for that test to be practical.

**Comment by peastman** [comment]
I looked into the bouncing motion you mentioned.  It looks to me like this line in `GodotCollisionSolver3D::solve_distance()` is the problem:

https://github.com/godotengine/godot/blob/57267709e665f1683e79bd5d3432be2be5db6c1d/servers/physics_3d/godot_collision_solver_3d.cpp#L587

Even though we've created analytic routines to identify collisions, it's still hardcoded to use GJK at a key point in the process.

File: servers/physics_3d/godot_collision_solver_3d_sat.cpp [hunk 1]
```diff
@@ -872,62 +857,50 @@ static void _collision_sphere_capsule(const GodotShape3D *p_a, const Transform3D
...
+	if (length > p_radius_a + p_margin_a + p_margin_b) {
 		return;
 	}
-
-	// Cylinder B end caps.
-	Vector3 cylinder_B_axis = p_transform_b.basis.get_column(1).normalized();
-	if (!separator.test_axis(cylinder_B_axis)) {
-		return;
+	p_collector->collided = true;
+	Vector3 axis;
```
Comment (Malcolmnixon) [inline servers/physics_3d/godot_collision_solver_3d_sat.cpp:884]: If the p_collector->callback is null you can bail here and save some computrons calculating the normalized axis as it won't be used if there is no callback. I just did the same change for my analytic sphere collision.

Comment (peastman) [reply servers/physics_3d/godot_collision_solver_3d_sat.cpp:888]: Good idea.  Done.

**Comment by Malcolmnixon** [comment]
By the looks of things we have two classes of collision algorithm:
 - solve_static detects when shapes are colliding, and returns the "contact points" on the shapes
 - solve_distance detects when shapes are not colliding, and returns the "closest points" on the shape

All the existing collision functions (_collision_sphere_sphere, _collision_sphere_box, _collision_capsule_capsule, etc.) are used by solve_static to detect the contact points and only return the points if the shapes are touching.

It seems like we're going to need a new class of functions (_distance_sphere_sphere, _distance_sphere_box, _distance_capsule_capsule, etc) to detect the closest points. These new functions would be VERY similar to the current analytic functions; but would provide their results on separation rather than collision, and wouldn't need to provide the points with a callback function.

**Comment by peastman** [comment]
I was just about to post asking for clarification about that.  It isn't clear to me whether we really need the distance version.  The only place `solve_distance()` ever gets called is inside `GodotSpace3D::test_body_motion()`, and that function appears to ignore all of the return values except the boolean for whether the objects overlap.

The way it gets called is rather confusing.  If you look at `solve_distance()`, you'll see that `r_point_A`, `r_point_B`, and `r_sep_axis` are strictly output arguments.  It stores values into them.  It doesn't care what initial values they have.  But `test_body_motion()` sets them anyway and seems to expect them to be used.  For example,

https://github.com/godotengine/godot/blob/61c7b7fb13f7cef8737d0a047db1c2340d1ce661/servers/physics_3d/godot_space_3d.cpp#L870-L871

Setting `sep` doesn't actually do anything at all.  The value just gets overwritten in the next line without ever being used.  It also records the values returned for the two nearest points

https://github.com/godotengine/godot/blob/61c7b7fb13f7cef8737d0a047db1c2340d1ce661/servers/physics_3d/godot_space_3d.cpp#L884-L885

but then they immediately get thrown out without ever being used.

It looks to me like `solve_distance()` is unnecessary.  All we really need is to check whether the two shapes overlap or not.

**Comment by Malcolmnixon** [comment]
I was thinking if we had an actual solve_distance() function we could use it like a ray-marcher for casting motion rather than the binary search approach used at the moment. I'm not certain how much of an improvement we would get in terms of the average number of iterations. At least the binary search algorithm is bounded in terms of iteration-count and distance error.

**Comment by peastman** [comment]
An easy way to get some of the same benefit is to call `project_range()` on each object to find the shortest and longest motion distances at which they could possibly intersect.  In many cases that should provide much tighter initial bounds for the search.

**Comment by peastman** [comment]
One other thought about it: if we can get accurate collision normals in all cases, we can use them to compute the derivative of the collision depth with respect to the motion distance.  That can be used to get faster convergence.

**Review by rburing — APPROVED** [review]

**Comment by rburing** [comment]
Very nice approach, and good discussion in the comments here about what can be done next. I'm working on cylinder-cylinder, so we'll be able to use an axis created by the closest points there too.

**Comment by akien-mga** [comment]
Thanks!

**Comment by cridenour** [comment]
Just to leave a note for anyone who comes here after upgrading to beta 12 (or 4.0 I guess) and has physics issues with spheres, **really make sure** that you don't have a collision shape scaled from itself or any parent nodes, as the previous solver would let you get away with it and this one will not.

It highlighted a bug in my code that I thought I had previously solved. Hopefully this helps someone in the future.

Also thank you @peastman and @Malcolmnixon for the focus on physics.

**Comment by realkotob** [comment]
@cridenour We should probably add a note in the documentation for this, and make an editor warning if a collision shape is scaled from itself or its parents.

**Pull request merged**